crates/aptos/src/account/multisig_account.rs

```diff
@@ -116,19 +130,102 @@ impl CliCommand<TransactionSummary> for CreateTransaction {
     }
 }
 
+/// Verify entry function matches on-chain transaction proposal.
+#[derive(Debug, Parser)]
+pub struct VerifyProposal {
+    #[clap(flatten)]
+    pub(crate) multisig_account_with_sequence_number: MultisigAccountWithSequenceNumber,
+    #[clap(flatten)]
+    pub(crate) txn_options: TransactionOptions,
+    #[clap(flatten)]
+    pub(crate) entry_function_args: EntryFunctionArguments,
+}
+
+#[async_trait]
+impl CliCommand<serde_json::Value> for VerifyProposal {
+    fn command_name(&self) -> &'static str {
+        "VerifyProposalMultisig"
+    }
+
+    async fn execute(self) -> CliTypedResult<serde_json::Value> {
+        // Get multisig transaction via view function.
+        let multisig_transaction = &self
+            .txn_options
+            .view(ViewRequest {
+                function: GET_TRANSACTION_ENTRY_FUNCTION.clone(),
+                type_arguments: vec![],
+                arguments: vec![
+                    serde_json::Value::String(String::from(
+                        &self
+                            .multisig_account_with_sequence_number
+                            .multisig_account
+                            .multisig_address,
+                    )),
+                    serde_json::Value::String(
+                        self.multisig_account_with_sequence_number
+                            .sequence_number
+                            .to_string(),
+                    ),
+                ],
+            })
+            .await?[0];
+        // Get expected multisig transaction payload hash hex from provided entry function.
+        let expected_payload_hash = HashValue::sha3_256_of(
+            &to_bytes::<MultisigTransactionPayload>(&self.entry_function_args.try_into()?)?,
+        )
+        .to_hex_literal();
+        // Get on-chain payload hash. If full payload provided on-chain:
+        let actual_payload_hash =
+            if let Some(actual_payload) = view_json_option_str(&multisig_transaction["payload"])? {
+                // Actual payload hash is the hash of the on-chain payload.
+                HashValue::sha3_256_of(actual_payload.parse::<HexEncodedBytes>()?.inner())
+                    .to_hex_literal()
+            // If full payload not provided, get payload hash directly from transaction proposal:
+            } else {
+                view_json_option_str(&multisig_transaction["payload_hash"])?.ok_or(
+                    CliError::UnexpectedError(
+                        "Neither payload nor payload hash provided on-chain".to_string(),
+                    ),
+                )?
+            };
+        // Get verification result based on if expected and actual payload hashes match.
+        let result = if expected_payload_hash.eq(&actual_payload_hash) {
+            json!({
+                "Status": "Transaction match",
+                "Multisig transaction": multisig_transaction
+            })
+        } else {
+            json!({
```

Comment:
Love this new JSON, much easier for the user to understand and know what to do with.

In this case will the CLI exit with success rather than failure? If so I wonder if there is a way we can enhance the CLI error type to support richer information via JSON, so that the CLI can exit with the expected return code.

Reply:
@banool presently the CLI returns with:

```bash
{
  "Result": {
    "Status": "Transaction mismatch",
    "Provided payload hash": "0xe494b0072d6f940317344967cf0e818c80082375833708c773b0275f3ad07e51",
    "On-chain payload hash": "0x070ed7c3f812f25f585461305d507b96a4e756f784e01c8c59901871267a1580",
    "For more information": "https://aptos.dev/move/move-on-aptos/cli#multisig-governance"
  }
}
```

I wouldn't necessarily classify this behavior as an error per se, since the inputs are well formed and the output is essentially a `bool` which is correctly evaluated (is or is not a match), but either way here is a snippet of what it might look like to force the CLI to exit with an error code:

```rust
        // Get verification result based on if expected and actual payload hashes match.
        if expected_payload_hash.eq(&actual_payload_hash) {
            Ok(json!({
                "Status": "Transaction match",
                "Multisig transaction": multisig_transaction
            }))
        } else {
            Err(CliError::UnexpectedError(
                TransactionMismatchErrorJSON::pretty_string(
                    expected_payload_hash,
                    actual_payload_hash,
                ),
            ))
        }
    }
}

#[derive(Deserialize, Serialize)]
struct TransactionMismatchErrorJSON {
    #[serde(rename = "Status")]
    status: &'static str,
    #[serde(rename = "Provided payload hash")]
    provided_hash: String,
    #[serde(rename = "On-chain payload hash")]
    on_chain_hash: String,
    #[serde(rename = "For more information")]
    for_more_info: &'static str,
}

impl TransactionMismatchErrorJSON {
    fn pretty_string(provided_hash: String, on_chain_hash: String) -> String {
        serde_json::to_string_pretty(&TransactionMismatchErrorJSON {
            status: "Transaction mismatch",
            provided_hash,
            on_chain_hash,
            for_more_info: "https://aptos.dev/move/move-on-aptos/cli#multisig-governance",
        })
        .expect("Unable to convert all-string JSON map to a pretty string")
    }
}
```

As you alluded to above the CLI error system isn't equipped to handle this output, and mangles the pretty JSON into a one-line string:

```bash
{
  "Error": "Unexpected error: {\n  \"Status\": \"Transaction mismatch\",\n  \"Provided payload hash\": \"0xe494b0072d6f940317344967cf0e818c80082375833708c773b0275f3ad07e51\",\n  \"On-chain payload hash\": \"0x070ed7c3f812f25f585461305d507b96a4e756f784e01c8c59901871267a1580\",\n  \"For more information\": \"https://aptos.dev/move/move-on-aptos/cli#multisig-governance\"\n}"
}
```

I don't see this as a major issue because the user is in some sense asking the CLI "does this match?" and the CLI is providing a correct response, but if you think that there should be a `CliError` exit then yes, it would appear there is substantial re-work required for the error system.

Your thoughts?

Reply:
Okay thanks for exploring this. While JSON-containing error messages are nice I don't think they're at all necessary. I'd say this approach is worth it, it is important for the CLI to error in the error case (though I agree it isn't clear cut whether this counts as an error). It can be useful for uses in CI / automation for example, as well as e2e testing.